contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1683 @@
+    /// @return Total weight of the node
+    function _getTotalBalancesIncludingWeight(address _node) internal view returns (uint256) {
+        Node memory node = nodes[_node];
+        return node.balance + node.sumOfLeftBalances + node.sumOfRightBalances;
+    }
+
+    /// @notice Returns the sum of balances of given nodes
+    /// @param _nodes List of nodes
+    /// @return Sum of balances of given nodes
+    function sumOfBalancesOfSelectedNodes(address[] memory _nodes) internal view returns (uint256) {
+        uint256 total;
+        for (uint256 index = 0; index < _nodes.length; index++) {
+            Node memory node = nodes[_nodes[index]];
+            total += node.balance;
+        }
+        return total;
+    }
+
+    // function _printNode(address _node) internal view {
+    //     Node memory node = nodes[_node];
+    //     console2.log("************************************");
+    //     console2.log("cluster", node.node);
+    //     console2.log("balance", node.balance);
+    //     console2.log("left", node.left);
+    //     console2.log("right", node.right);
+    //     console2.log("sumOfLeftBalances", node.sumOfLeftBalances);
+    //     console2.log("sumOfRightBalances", node.sumOfRightBalances);
+    //     console2.log(" height", node.height);
+    //     console2.log("************************************");
+    // }
+
+    // function _printArray(string memory data, bytes memory arrayBytes) internal view {
+    //     console2.log(data);
+    //     address[] memory array = abi.decode(arrayBytes, (address[]));
+    //     console2.log("[");
+    //     for (uint256 index = 0; index < array.length; index++) {
+    //         console2.log(index, array[index]);
+    //     }
+    //     console2.log("]");
+    // }
+
+    // function _printArray(string memory data, address[] memory array) internal view {
+    //     console2.log(data);
+    //     console2.log("[");
+    //     for (uint256 index = 0; index < array.length; index++) {
+    //         console2.log(index, array[index]);
+    //     }
+    //     console2.log("]");
+    // }
+
+    // function _printPaths(string memory data, bytes[] memory bytesdata) internal view {
+    //     console2.log(data);
+
+    //     console2.log("[");
+    //     for (uint256 index = 0; index < bytesdata.length; index++) {
+    //         address[] memory _paths = abi.decode(bytesdata[index], (address[]));
+    //         _printArray("subarray ", _paths);
+    //     }
+    //     console2.log("]");
+    // }
+}
+
+contract SingleSelector is SelectorHelper {
+    constructor() {}
+
+    /// @inheritdoc IClusterSelector
+    function insert(address newNode, uint256 balance) public virtual override {
+        require(newNode != address(0), ClusterLib.CANNOT_BE_ADDRESS_ZERO);
+        Node memory node = nodes[newNode];
+        if (node.node == address(0)) {
+            root = _insert(root, newNode, balance);
+            totalElements++;
```

Comment:
Do we need `totalElements` variable ? It seems to be used only to check if there are atleast N elements in the tree when selecting topNNodes. Can't we find that if the weight of root after excluding the selected nodes is becoming 0 before we find all top N elements.

Reply:
It is a small cost but might be useful in case we want to extend the contracts at a later point.